README.md
# OssaNMA

This is a package for the optimal sample size allocation for a new trial when leveraging evidence from existing network meta-analysis

## How to install
You can install this through use devtools

`devtools::install_github("fangshuye/OssaNMA")`

## Import
`library(OssaNMA)`


NEWS.md
# OssaNMA 0.1.2

- Change an example in the vignette
- Add power value to the output of ssnma()

---

# OssaNMA 0.1.1

- Change the rounding to be rounding up any decimal value to the next integer value for the optimal sample size

---

# OssaNMA 0.1.0

- new package release: OssaNMA


R/ssnma.R
#' @importFrom stats pnorm qnorm
NULL

#' Calculate the optimal sample sizes for a new two-arm trial when analyze it with the existing network
#'
#' This function calculates the optimal sample size for each treatment group to
#' achieve a pre-specified power when planning a new two-arm trial with
#' binary outcome.
#'
#' @param p1 Risk of treatment 1
#' @param p2 Risk of treatment 2
#' @param enma_sigma Standard error of the estimated effect size (log odds ratio) between treatment 1 and treatment 2 from the existing network
#' @param power.level Power of test we want to obtain
#' @param sig.level Significance level, the default value is 0.05
#' @param method a character string specifying the method of analyzing the new trial, must be one of 'with' (default) or 'without'
#' @param allocation a character string specifying the type of sample size allocation between two groups, must be one of 'uneven' (default) or 'even'.
#' @return A list with the following components:
#'  \item{sample_size}{Sample size for each treatment group.}
#'  \item{power}{Power of the test.}
#' @export
#' @examples
#' ssnma(p1 = 0.2, p2 = 0.3, enma_sigma = 0.4, power = 0.8)

ssnma <- function(p1,p2,enma_sigma = NULL,power.level,sig.level = 0.05, method = "with", allocation = "uneven"){
  if( any(p1 < 0 | p1 > 1) ) stop('p1 not between 0 and 1')
  if( any(p2 < 0 | p2 > 1) ) stop('p2 not between 0 and 1')
  if( any(power.level < 0 | power.level > 1) ) stop('power.level not between 0 and 1')
  if( any(sig.level < 0 | sig.level > 1) ) stop('sig.level not between 0 and 1')

  if(allocation == "uneven"){
    SSS_uneven(p1,p2,enma_sigma,power.level,sig.level, method)
  }else{
    SSS_even(p1,p2,enma_sigma,power.level,sig.level, method)
  }

}


R/BRDdat.R
#' Example dataset for network meta-analysis
#'
#' This example dataset represents a previously published network of interventions
#' for the treatment of Bovine Respiratory Disease (BRD) in feedlot cattle
#' (O’Connor, Yuan, Cullen, Coetzee, Da Silva, and Wang, 2016). The dataset is
#' comprised of 98 trials, 13 treatments and 204 arms. Each row represents the
#' summary statistics for a pairwise comparison between two treatment in a trial.
#'
#' @docType data
#'
#' @usage data(BRDdat)
#'
#' @format An object of class \code{"data.frame"}
#' \describe{
#'  \item{studlab}{study id (integer)}
#'  \item{treat1}{name of treatment 1 (character)}
#'  \item{treat2}{name of treatment 2 (character)}
#'  \item{TE}{estimated treatment effect size (log odds ratio) between
#'    treat1 and treat2}
#'  \item{seTE}{standard error of TE}
#' }
#' @references \href{https://pubmed.ncbi.nlm.nih.gov/27612392/}{O’Connor, A. M., Yuan, C., Cullen, J. N., Coetzee, J. F., Da Silva, N., & Wang, C. (2016). A mixed treatment meta-analysis of antibiotic treatment options for bovine respiratory disease–an update. Preventive veterinary medicine, 132, 130-139.}.
#' @keywords datasets
#' @examples
#'
#' data(BRDdat)
#' head(BRDdat)
#'
"BRDdat"


R/SSS_equal.R
SSS_even <- function(p1,p2,enma_sigma,power.level,sig.level = 0.05, method = "with"){
  # cal the total sample size when we added one more condition: sample sizes are equal for both treatment groups
  beta1 <- log(p1/(1-p1))
  beta2 <- log(p2/(1-p2)) - log(p1/(1-p1))
  mu_1 <- exp(beta1)/(1+exp(beta1))^2
  mu_2 <- exp(beta1 + beta2)/(1+exp(beta1 + beta2))^2
  n0 <- c(1,1)


  power_cal <- function(n){
    size <- n/2
    if(method == "with") {
      if( is.null(enma_sigma) ) stop('enma_sigma is missing')
      var_inv <- 1/(1/(mu_1 * size) + 1/(mu_2 * size))+1/enma_sigma^2
      var <- 1/var_inv
    }else{
      var <- 1/(mu_1 * size) + 1/(mu_2 * size)
    }
    se <- sqrt(var)
    z <- beta2/se
    power <- pnorm(z-qnorm(1-sig.level/2))+pnorm(-z-qnorm(1-sig.level/2))
    return(power)
  }

  confun <- function(n){
    f = power.level-power_cal(n)
    f = rbind(f,-n)
    return(list(ceq=NULL,c=f))
  }

  objfun=function(n){
    n[1]+n[2]
  }

  solution_temp <- NlcOptim::solnl(n0,objfun=objfun,confun=confun)$par
  solution_temp_int <- ceiling(solution_temp[2])
  # get the even integer solution around this
  if(solution_temp_int %% 2 ==0){
    res <- solution_temp_int
  }else{
    res <- solution_temp_int+1
  }

  power_value <- round(power_cal(res),3)
  return(list(sample_size = c(res/2,res/2),
              power = power_value))

}



R/SSS_unequal.R
SSS_uneven <- function(p1,p2,enma_sigma,power.level,sig.level = 0.05, method = "with"){
  beta1 <- log(p1/(1-p1))
  beta2 <- log(p2/(1-p2)) - log(p1/(1-p1))
  mu_1 <- exp(beta1)/(1+exp(beta1))^2
  mu_2 <- exp(beta1 + beta2)/(1+exp(beta1 + beta2))^2
  n0 <- c(1,1)

  power_cal <- function(n){
    if(method == "with"){
      if( is.null(enma_sigma) ) stop('enma_sigma is missing')
      var_inv <- 1/(1/(mu_1 * n[1]) + 1/(mu_2 * n[2]))+1/enma_sigma^2
      var <- 1/var_inv
    }else{
      var <- 1/(mu_1 * n[1]) + 1/(mu_2 * n[2])
    }

    se <- sqrt(var)
    z <- beta2/se
    power <- pnorm(z-qnorm(1-sig.level/2))+pnorm(-z-qnorm(1-sig.level/2))
    return(power)

  }

  confun <- function(n){
    f = power.level-power_cal(n)
    f = rbind(f,-n[1])
    f = rbind(f,-n[2])
    return(list(ceq=NULL,c=f))
  }


  objfun=function(n){
    n[1]+n[2]
  }

  solution_temp <- NlcOptim::solnl(n0,objfun=objfun,confun=confun)$par
  solution_integer <- ceiling(solution_temp)

  n_each_group <- solution_integer[,1]
  power_value <- round(power_cal(n_each_group),3)
  return(list(sample_size = n_each_group,
              power = power_value))

}


R/ssanma.R
#' @importFrom stats pnorm qnorm
NULL

#' Calculate the optimal sample size allocation for a new two-arm trial when analyze it with the existing network
#'
#' This function calculates the optimal sample size allocation for each treatment group
#' with a fixed total sample size when planning a new two-arm trial
#' with binary outcome.
#'
#' @param p1 Risk of treatment 1
#' @param p2 Risk of treatment 2
#' @param enma_sigma Standard error of the estimated effect size (log odds ratio) between treatment 1 and treatment 2 from the existing network
#' @param N Number of total sample size
#' @param sig.level Significance level, the default value is 0.05
#' @param method a character string specifying the method of analyzing the new trial, must be one of 'with' (default) or 'without'
#' @param allocation a character string specifying the type of sample size allocation between two groups, must be one of 'uneven' (default) or 'even'.
#' @return A list with the following components:
#'  \item{sample_alloc}{Sample size allocation to each treatment group.}
#'  \item{power}{Power of the test.}
#' @export
#' @examples
#' ssanma(p1 = 0.2, p2 = 0.3, enma_sigma = 0.4, N = 200)
ssanma <- function(p1,p2,enma_sigma,N,sig.level = 0.05, method = "with", allocation = "uneven"){
  # cal the total sample size when we added one more condition: sample sizes are equal for both treatment groups
  beta1 = log(p1/(1-p1))
  beta2 = log(p2/(1-p2)) - log(p1/(1-p1))
  mu_1 <- exp(beta1)/(1+exp(beta1))^2
  mu_2 <- exp(beta1 + beta2)/(1+exp(beta1 + beta2))^2
  s <- N
  var_fn <- function(n, mu_1, mu_2, s){
    1/(mu_1 * n[1]) + 1/(mu_2 * n[2])
  }

  power_formula <- function(n){
    if(method == "with") {
      if( is.null(enma_sigma) ) stop('enma_sigma is missing')
      var_inv <- 1/(1/(mu_1 * n[1]) + 1/(mu_2 * n[2]))+1/enma_sigma^2
      var <- 1/var_inv
    }else{
      var <- 1/(mu_1 * n[1]) + 1/(mu_2 * n[2])
    }
    se <- sqrt(var)
    z <- beta2/se
    power <- pnorm(z-qnorm(1-sig.level/2))+pnorm(-z-qnorm(1-sig.level/2))
    return(power)
  }

  con_fn <- function(n, mu_1, mu_2,s){
    n1 <- n[1]
    n2 <- n[2]
    n1 + n2 - s
  }
  if(allocation == "uneven"){
    suppressWarnings(sample_alloc <- DEoptimR::JDEoptim(lower = c(0,0), upper = c(s,s),
                                       fn = var_fn, constr = con_fn,
                                       meq = 1, triter = 50,
                                       mu_1 = mu_1,
                                       mu_2 = mu_2,
                                       s = s))
    sample_alloc <- round(sample_alloc$par,0)
  }else{
    sample_alloc <- c(N/2, N/2)
  }


  power_value <- round(power_formula(sample_alloc),3)
  return(list(sample_alloc = sample_alloc, power = power_value))
}

